pandas/core/internals/blocks.py

```diff
@@ -860,14 +864,21 @@ def setitem(self, indexer, value):
             #  be e.g. a list; see GH#6043
             values[indexer] = value
 
-        # if we are an exact match (ex-broadcasting),
-        # then use the resultant dtype
         elif (
-            len(arr_value.shape)
-            and arr_value.shape[0] == values.shape[0]
-            and arr_value.size == values.size
+            is_categorical_dtype(arr_value.dtype)
```

Comment:
can you put exact_match as the first condition

Reply:
@jreback moved exact_match to be the first condition, and merged master, all tests passed.